mlflow/models/evaluation/default_evaluator.py

```diff
@@ -0,0 +1,405 @@
+            if y_probs is not None:
+                fpr, tpr, thresholds = sk_metrics.roc_curve(y, y_prob)
+                roc_curve_pandas_df = pd.DataFrame(
+                    {"fpr": fpr, "tpr": tpr, "thresholds": thresholds}
+                )
+                self._log_pandas_df_artifact(
+                    artifacts, temp_dir, roc_curve_pandas_df, run_id, "roc_curve", dataset_name, model,
+                )
+
+                roc_auc = sk_metrics.auc(fpr, tpr)
+                metrics["roc_auc"] = roc_auc
+
+                def plot_roc_curve():
+                    sk_metrics.RocCurveDisplay(
+                        fpr=fpr, tpr=tpr, roc_auc=roc_auc,
+                    ).plot()
+
+                if hasattr(sk_metrics, 'RocCurveDisplay'):
+                    self._log_image_artifact(
+                        artifacts, temp_dir, plot_roc_curve, run_id, "roc_curve", dataset_name,
+                    )
+
+                precision, recall, thresholds = sk_metrics.precision_recall_curve(y, y_prob)
+                thresholds = np.append(thresholds, [1.0], axis=0)
+                pr_curve_pandas_df = pd.DataFrame(
+                    {"precision": precision, "recall": recall, "thresholds": thresholds}
+                )
+                self._log_pandas_df_artifact(
+                    artifacts, temp_dir, pr_curve_pandas_df, run_id, "precision_recall_curve",
+                    dataset_name, model,
+                )
+
+                pr_auc = sk_metrics.auc(recall, precision)
+                metrics["precision_recall_auc"] = pr_auc
+
+                def plot_pr_curve():
+                    sk_metrics.PrecisionRecallDisplay(
+                        precision, recall,
+                    ).plot()
+
+                if hasattr(sk_metrics, 'PrecisionRecallDisplay'):
+                    self._log_image_artifact(
+                        artifacts, temp_dir, plot_pr_curve, run_id, "precision_recall_curve", dataset_name,
+                    )
+
+                self._log_image_artifact(
+                    artifacts, temp_dir,
+                    lambda: plot_lift_curve(y, y_probs),
+                    run_id, "lift_curve", dataset_name,
+                )
+
+            def plot_confusion_matrix():
+                sk_metrics.ConfusionMatrixDisplay(confusion_matrix=confusion_matrix).plot()
+
+            if hasattr(sk_metrics, 'ConfusionMatrixDisplay'):
+                self._log_image_artifact(
+                    artifacts, temp_dir, plot_confusion_matrix, run_id, "confusion_matrix", dataset_name,
+                )
+
+        self._log_metrics(run_id, metrics, dataset_name)
+
+        if evaluator_config.get('log_model_explainability', True):
+            self._log_model_explainability(
+                artifacts, temp_dir, model, X, dataset_name, feature_names, run_id, evaluator_config
+            )
```

Comment:
I ran the following code:

```
import mlflow
from mlflow.models.evaluation import EvaluationDataset

from sklearn import datasets
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

# prepare train and test data
iris = datasets.load_iris()
X = iris.data
y = iris.target
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

with mlflow.start_run():
    rid = mlflow.active_run().info.run_id

    logistic = LogisticRegression(C=1.0)
    logistic.fit(X_train, y_train)
    mlflow.sklearn.log_model(logistic, artifact_path="mymodel")

    results = mlflow.evaluate(
        model="runs:/%s/mymodel" % rid,
        model_type="classifier",
        dataset=EvaluationDataset(X_test, y_test),
        evaluator_config={
            "log_model_explainability": False,
        },
        evaluators="default",
    )
    print(results)
```

This failed with:

```
Unable to serialize underlying model using MLflow, will use SHAP serialization
Traceback (most recent call last):
  File "eval.py", line 21, in <module>
    results = mlflow.evaluate(
  File "/Users/corey.zumar/mlflow/mlflow/models/evaluation/base.py", line 642, in evaluate
    return _evaluate(
  File "/Users/corey.zumar/mlflow/mlflow/models/evaluation/base.py", line 563, in _evaluate
    result = evaluator.evaluate(model, model_type, dataset, actual_run_id, config)
  File "/Users/corey.zumar/mlflow/mlflow/models/evaluation/default_evaluator.py", line 399, in evaluate
    return self._evaluate_classifier(
  File "/Users/corey.zumar/mlflow/mlflow/models/evaluation/default_evaluator.py", line 358, in _evaluate_classifier
    self._log_model_explainability(
  File "/Users/corey.zumar/mlflow/mlflow/models/evaluation/default_evaluator.py", line 216, in _log_model_explainability
    mlflow.shap.log_explainer(
  File "/Users/corey.zumar/mlflow/mlflow/shap.py", line 340, in log_explainer
    Model.log(
  File "/Users/corey.zumar/mlflow/mlflow/models/model.py", line 190, in log
    flavor.save_model(path=local_path, mlflow_model=mlflow_model, **kwargs)
  File "/Users/corey.zumar/mlflow/mlflow/shap.py", line 452, in save_explainer
    explainer.save(explainer_output_file_handle)
  File "/Users/corey.zumar/opt/anaconda3/lib/python3.8/site-packages/shap/explainers/_explainer.py", line 407, in save
    s.save("model", self.model, model_saver)
  File "/Users/corey.zumar/opt/anaconda3/lib/python3.8/site-packages/shap/_serializable.py", line 87, in save
    if len(inspect.getfullargspec(value.save)[0]) == 3: # backward compat for MLflow, can remove 4/1/2021
AttributeError: 'LogisticRegression' object has no attribute 'save'
```

Which is surprising because I specified `log_model_explainability`: False. I printed out the evaluator config above this line and found that its empty. We must be losing it somewhere. Can we fix this and start writing some test cases to cover these common behaviors?

Reply:
Fixed in the latest code. Maybe previous code I miswrite 'log_model_explainability' key in code